nexus/db-queries/src/db/queries/region_allocation.rs

```diff
@@ -348,30 +355,61 @@ impl DoInsert {
             .single_value()
             .assume_not_null()
             .lt(redundancy);
-        let enough_candidates = candidate_regions
+
+        let enough_candidate_zpools = candidate_zpools
             .query_source()
             .count()
             .single_value()
             .assume_not_null()
             .ge(redundancy);
-        let proposals_fit = proposed_datasets_fit
+
+        let enough_candidate_regions = candidate_regions
             .query_source()
-            .select(bool_and(proposed_datasets_fit::dsl::fits))
+            .count()
             .single_value()
-            .assume_not_null();
+            .assume_not_null()
+            .ge(redundancy);
+
+        // We want to ensure that we do not allocate two regions on the same
+        // zpool, for two reasons
+        // - Data redundancy: If a drive fails it should only take one of the 3
+        //   regions with it
+        // - Risk of overallocation: We only check that each zpool as enough
+        //   room for one region, so we should not allocate more than one region
+        //   to it.
+        //
+        // If we have enough candidate zpools, but we allocated two regions on
+        // the same pool, then the code running the query will catch this error
+        // and try again to see if a different dataset shuffle works out.
+        use crate::db::schema::dataset::dsl as dataset_dsl;
+        use candidate_regions::dsl as candidate_dsl;
+        let enough_unique_candidate_zpools = candidate_regions
+            .query_source()
+            .inner_join(
+                dataset_dsl::dataset
+                    .on(candidate_dsl::dataset_id.eq(dataset_dsl::id)),
+            )
+            .select(diesel::dsl::count_distinct(dataset_dsl::pool_id))
+            .single_value()
+            .assume_not_null()
+            .ge(redundancy);
```

Comment:
See my comment above -- I think ideally we would make this check when selecting candidate datasets. Performing it as an after-the-fact check means we could potentially "pick wrong" several times, even when we could otherwise know the right selections of datasets to avoid this problem ahead of time.